src/cpp/session/SessionSourceDatabaseSupervisor.cpp

```diff
@@ -116,8 +116,21 @@ FilePath sessionLockFilePath(const FilePath& sessionDir)
    return sessionDir.complete("lock_file");
 }
 
-// session dir lock (initialized by attachToSourceDatabase)
-FileLock s_sessionDirLock;
+// session dir lock (lock is acquired within 'attachToSourceDatabase()')
+boost::shared_ptr<FileLock> createSessionDirLock()
+{
+#ifdef RSTUDIO_SERVER
+   return FileLock::create(FileLock::LOCKTYPE_LINKBASED);
+#else
+   return FileLock::createDefault();
+#endif
+}
+
+boost::shared_ptr<FileLock>& sessionDirLock()
+{
+   static boost::shared_ptr<FileLock> instance = createSessionDirLock();
```

Comment:
This should be safe (the destructor for the shared_ptr will be called when the program exits; this implies calling `delete` on the associated object and hence its destructor).

I can verify this in some tests.